app/questionnaire/questionnaire_manager.py

```diff
@@ -1,38 +1,242 @@
-import bleach
+from app.questionnaire.state_manager import StateManager
+from app.questionnaire_state.introduction import Introduction as StateIntroduction
+from app.questionnaire_state.page import Page
+from app.questionnaire_state.summary import Summary as StateSummary
+from app.questionnaire_state.thank_you import ThankYou as StateThankYou
+from app.schema.questionnaire import QuestionnaireException
+from app.metadata.metadata_store import MetaDataStore
 from app.templating.template_pre_processor import TemplatePreProcessor
-
+from app.routing.routing_engine import RoutingEngine
 from app.questionnaire.user_action_processor import UserActionProcessor
 from app.authentication.session_management import session_manager
 from flask_login import current_user
 import logging
+import bleach
 
 
 logger = logging.getLogger(__name__)
 
 
 class QuestionnaireManager(object):
-    def __init__(self, schema, user_journey_manager, routing_engine, metadata):
+    '''
+    This class represents a user journey through a survey. It models the request/response process of the web application
+    using a doubly linked list. Each node in the list is essentially a page displayed to the user. A new page is created
+    by a GET request and subsequently updated via a POST request.
+
+    The doubly linked list approach allows us to maintain the path the user has taken through the question. If that path
+    changes we archive off the pages incase the user revists that path.
+
+    '''
+    def __init__(self, schema):
+        self.submitted_at = None
         self._schema = schema
-        self._user_journey_manager = user_journey_manager
-        self._metadata = metadata
-        self._routing_engine = routing_engine
-        self._user_action_processor = UserActionProcessor(self._schema, self._metadata, self._user_journey_manager)
-        self._pre_processor = TemplatePreProcessor(self._schema, self._user_journey_manager, self._metadata)
+        self._current = None  # the latest page
+        self._first = None  # the first page in the doubly linked list
+        self._archive = {}  # a dict of completed or discarded pages
+        self._valid_locations = self._build_valid_locations()
+        self._pre_processor = TemplatePreProcessor(self._schema, self)
+
+    @staticmethod
+    def new_instance(schema):
```

Comment:
new_instance and get_instance is a standard singleton pattern though. It's just using the DB to maintain the instance rather than a variable. I'm reluctant to change this one to be honest.